skxray/cdi.py

```diff
@@ -119,9 +142,284 @@ def convolution(array1, array2):
+
+        # initial support
+        self.init_obj_flag = kwargs['init_obj_flag']
+        self.init_sup_flag = kwargs['init_sup_flag']
+        self.sup_radius = kwargs['support_radius']
+        self.sup_shape = kwargs['support_shape']  # 'Box' or 'Disk'
+
+        # parameters for shrink wrap
+        self.shrink_wrap_flag = kwargs['shrink_wrap_flag']
+        self.sw_sigma = kwargs['sw_sigma']  # 0.5
+        self.sw_threshold = kwargs['sw_threshold']  # 0.1
+        self.sw_start = kwargs['sw_start']  # 0.2
+        self.sw_end = kwargs['sw_end']  # 0.8
+        self.sw_step = kwargs['sw_step']  # 10
+
+    def init_obj(self):
+        """
+        Initiate phase. Focus on 2D case here, and to be extended.
+        """
+        pha_tmp = np.random.uniform(0, 2*np.pi, (self.nx, self.ny))
+        self.obj = (_fft_helper(self.diff_array * np.exp(1j*pha_tmp), option='ifft')
+                    * np.sqrt(np.size(self.diff_array)))
+
+    def init_sup(self):
+        """ Initiate sample support.
+        """
+        self.sup = np.zeros_like(self.diff_array)
+        if self.ndim == 2:
+            if self.sup_shape == 'Box':
+                self.sup[self.nx/2-self.sup_radius: self.nx/2+self.sup_radius,
+                         self.ny/2-self.sup_radius:self.ny/2+self.sup_radius] = 1
+        if self.ndim == 3:
+            if self.sup_shape == 'Box':
+                self.sup[self.nx/2-self.sup_radius:self.nx/2+self.sup_radius,
+                         self.ny/2-self.sup_radius:self.ny/2+self.sup_radius,
+                         self.nz/2-self.sup_radius:self.nz/2+self.sup_radius] = 1
+        if self.sup_shape == 'Disk':
+            dummy = _dist(np.shape(self.diff_array))
+            self.sup_index = np.where(dummy <= self.sup_radius)
+            self.sup[self.sup_index] = 1
+
+        self.sup_index = np.where(self.sup == 1)
+        self.sup_out_index = np.where(self.sup == 0)
+
+    def recon(self, n_iterations=1000):
+        """
+        Run reconstruction with difference map algorithm.
+
+        Parameters
+        ---------
+        n_iterations : int
+            number of reconstructions to run
+
+        Returns
+        -------
+        array :
+            reconstructed sample object
+        """
+        gamma_1 = -1/self.beta
+        gamma_2 = 1/self.beta
+
+        # initiate shape and phase
+        if(self.init_obj_flag):
+            self.init_obj()
+        if(self.init_sup_flag):
+            self.init_sup()
+
+        self.obj_error = np.zeros(n_iterations)
+        self.diff_error = np.zeros(n_iterations)
+        self.sup_error = np.zeros(n_iterations)
+
+        sup_old = np.zeros_like(self.diff_array)
+        obj_ave = np.zeros_like(self.diff_array).astype(complex)
+        ave_i = 0
+
+        self.time_start = time.time()
+        for n in range(n_iterations):
+            self.obj_old = np.array(self.obj)
+
+            obj_a = pi_modulus(self.obj, self.diff_array, self.pi_modulus_flag)
```

Comment:
`obj` is usally the abbreviation for 'object' when writing generic code that can take in _any_ python object, it is really confusing to use it here.